package/libskylark/src/libskylark.c

```diff
@@ -0,0 +1,386 @@
+    case E_SUB_CONNECTION_ERROR:
+      return "Client: Subscribe connection failed";
+    case E_SUB_WRITE_ERROR:
+      return "Client: Error writing SBP to Piksi";
+    case E_PUB_CONNECTION_ERROR:
+      return "Client: Publish connection failed";
+    case E_PUB_READ_ERROR:
+      return "Client: Error reading SBP from Piksi";
+    case E_BAD_HTTP_REQUEST:
+      return "Server: Bad HTTP header or request";
+    case E_NO_ROVER_POS_FOUND:
+      return "Server: Required rover position was not found";
+    case E_UNAUTHORIZED_CLIENT:
+      return "Server: Client is not authorized to use this service";
+    case E_RETRIES_EXHAUSTED:
+      return "Client: Client connection retries exhausted";
+    case E_MAX_ERROR:
+    default: {
+      return "Client: UNHANDLED ERROR";
+    }
+  }
+}
+
+/** Debug log client return code.
+ *
+ * \param code  Client return code
+ */
+void log_client_error(RC code)
+{
+  log_debug("client_debug code=%d msg=%s\n", code, client_strerror(-code));
+}
+
+/** Debug log CURL error codes
+ *
+ * \param code  curl return code
+ */
+static void log_curl_error(CURLcode code)
+{
+  log_error("curl_debug code=%d msg=%s\n", code, curl_easy_strerror(code));
+}
+
+/**
+ *  Configuration
+ */
+
+/** Debug log a human-readable representation of the Skylark configuration.
+ *
+ * \param config  Skylark client configuration.
+ * \return  RC
+ */
+void log_client_config(const client_config_t *config)
+{
+  log_debug("client_config: endpoint_url=%s\n", config->endpoint_url);
+  log_debug("client_config: device_uuid=%s\n", config->device_uuid);
+  log_debug("client_config: fd=%d\n", config->fd);
+  log_debug("client_config: num_retries=%d\n", config->num_retries);
+  log_debug("client_config: retry_delay=%d\n", config->retry_delay);
+  log_debug("client_config: retry_max_time=%d\n", config->retry_max_time);
+}
+
+/** Read the Skylark device UUID.
+ *
+ * \param uuid  UUID4 string to return.
+ * \return  RC return code indicating success or failure
+ */
+RC get_device_uuid(char *uuid)
+{
+  if (file_read_string(FILE_SKYLARK_UUID, uuid, UUID4_SIZE) < 0) {
+    return E_GENERIC_ERROR;
+  }
+  return NO_ERROR;
+}
+
+/** Format a device UUID (UUID4) for an HTTP request.
+ *
+ * \param uuid         Device UUID4
+ * \param uuid_header  Formatted string header to return.
+ * \return  RC return code indicating success or failure
+ */
+RC get_device_header(const char *uuid, char *uuid_header)
+{
+  // TODO: Handle error code here
+  sprintf(uuid_header, DEVICE_UID_HEADER_FMT, uuid);
+  return NO_ERROR;
+}
+
+/** Initialize a Skylark client configuration.
+ *
+ * \param config Client configuration to initialize
+ * \return  RC return code indicating success or failure
+ */
+RC init_config(client_config_t *config)
+{
+  RC rc = NO_ERROR;
+  if ((rc = get_device_uuid(config->device_uuid)) < NO_ERROR) {
+    return rc;
+  }
+  if ((rc = get_device_header(config->device_uuid, config->device_header)) <
```

Comment:
@mfine tried to materialize header string this within the upload/download processes but started getting bad HTTP header requests that I didn't understand.